powersimdata/network/europe_tub/model.py

```diff
@@ -0,0 +1,140 @@
+
+    def retrieve_data(self):
+        """Fetch data"""
+        zenodo_get.zenodo_get([self.zenodo_record_id, "-o", f"{self.data_loc}"])
+        with ZipFile(os.path.join(self.data_loc, "networks.zip"), "r") as zip_network:
+            zip_network.extractall(self.data_loc)
+
+    def build(self):
+        """Build network"""
+        path = os.path.join(self.data_loc, "networks")
+        if self.reduction is None:
+            network = pypsa.Network(os.path.join(path, "elec_s.nc"))
+        elif os.path.exists(os.path.join(path, f"elec_s_{self.reduction}.nc")):
+            network = pypsa.Network(os.path.join(path, f"elec_s_{self.reduction}.nc"))
+        else:
+            raise ValueError(
+                "Invalid Resolution. Choose among: None | 1024 | 512 | 256 | 128 | 37"
+            )
+        id2zone = {i: l for i, l in enumerate(network.buses.index)}
+        zone2id = {l: i for i, l in id2zone.items()}
+
+        if "Europe" in self.interconnect:
+            self.network = network
+            self.id2zone = id2zone
+            self.zone2id = zone2id
+        else:
+            filter = list(  # noqa: F841
+                interconnect2abv[
+                    interconnect_to_name(self.interconnect, model=self.grid_model)
+                ]
+            )
+            self.network = network[network.buses.query("country == @filter").index]
+            self.zone2id = {l: zone2id[l] for l in self.network.buses.index}
+            self.id2zone = {i: l for l, i in self.zone2id.items()}
+
+        self.model_immutables = self._generate_model_immutables()
+
+    def _generate_model_immutables(self):
+        """Generate the model immutables"""
+        mapping = ModelImmutables(self.grid_model, interconnect=self.interconnect)
+
+        # loadzone
+        mapping.zones["loadzone"] = set(self.zone2id)
+        mapping.zones["id2loadzone"] = self.id2zone
+        mapping.zones["loadzone2id"] = self.zone2id
+        mapping.zones["loadzone2abv"] = self.network.buses["country"].to_dict()
+        mapping.zones["loadzone2country"] = {
+            l: abv2country[a]
+            for l, a in self.network.buses["country"].to_dict().items()
+        }
+        mapping.zones["loadzone2interconnect"] = {
+            l: mapping.zones["abv2interconnect"][a]
+            for l, a in mapping.zones["loadzone2abv"].items()
+        }
+        mapping.zones["id2timezone"] = {
+            self.zone2id[l]: abv2timezone[a]
+            for l, a in mapping.zones["loadzone2abv"].items()
+        }
+        mapping.zones["timezone2id"] = {
+            t: i for i, t in mapping.zones["id2timezone"].items()
+        }
+
+        # country
+        mapping.zones["country2loadzone"] = {
+            abv2country[a]: set(l)
+            for a, l in self.network.buses.groupby("country").groups.items()
+        }
+        mapping.zones["abv2loadzone"] = {
+            a: set(l) for a, l in self.network.buses.groupby("country").groups.items()
+        }
```

Comment:
Is this equivalent to
```python
mapping.zones["abv2loadzone"] = (
    self.network.buses.groupby("country").groups.map(set).to_dict()
)
```
?

Reply:
This gives me an `AttributeError`:
```
AttributeError: 'PrettyDict' object has no attribute 'map'
```

Reply:
Ah, right, `.groups` is a `PrettyDict`, not a pandas `Series`.

Reply:
I have kept the current implementation